azure-iot-device/tests/common/pipeline/pipeline_stage_test.py

```diff
@@ -0,0 +1,167 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import pytest
+from tests.common.pipeline.helpers import (
+    all_except,
+    make_mock_stage,
+    make_mock_op_or_event,
+    assert_callback_failed,
+    UnhandledException,
+)
+from tests.common.pipeline.pipeline_data_object_test import add_instantiation_test
+
+logging.basicConfig(level=logging.INFO)
+
+
+def add_base_pipeline_stage_tests(cls, module, all_ops, handled_ops, all_events, handled_events):
+    """
+    Add all of the "basic" tests for validating a pipeline stage.  This includes tests for
+    instantiation and tests for properly handling "unhandled" operations and events".
+    """
+
+    add_instantiation_test(
+        cls=cls,
+        module=module,
+        defaults={"name": cls.__name__, "next": None, "previous": None, "pipeline_root": None},
+    )
+    add_unknown_ops_tests(cls=cls, module=module, all_ops=all_ops, handled_ops=handled_ops)
+    add_unknown_events_tests(
+        cls=cls, module=module, all_events=all_events, handled_events=handled_events
+    )
+
+
+def add_unknown_ops_tests(cls, module, all_ops, handled_ops):
+    """
+    Add tests for properly handling of "unkonwn operations," which are operations that aren't
+    handled by a particular stage.  These operations should be passed down by any stage into
+    the stages that follow.
+    """
+    unknown_ops = all_except(all_items=all_ops, items_to_exclude=handled_ops)
+
+    @pytest.mark.describe("{} - .run_op() -- unknown and unhandled operations".format(cls.__name__))
+    @pytest.mark.parametrize("op_cls", unknown_ops)
+    class LocalTestObject(object):
+        @pytest.fixture
+        def op(self, op_cls, callback):
+            op = make_mock_op_or_event(op_cls)
+            op.callback = callback
+            op.action = "pend"
+            return op
+
+        @pytest.fixture
+        def stage(self, mocker):
+            return make_mock_stage(mocker=mocker, stage_to_make=cls)
+
+        @pytest.mark.it("Passes unknown operation to next stage")
```

Comment:
I'm not quite clear on the parametrization here. Output looks something like

```CoordinateRequestAndResponse - .run_op() -- unknown and unhandled operations
[x] Passes unknown operation to next stage[Connect]
...```

In this example, `Connect` would be the next stage right?

Why do we need to test this differently for each possible type of next stage? Shouldn't the next stage be somewhat of a black box from the perspective of the unit under test, in this case, the `CoordinateRequestAndResponse` stage?

I know we briefly talked about this, and it made sense to me at the time, but I'm a little confused looking at it now.

Reply:
I like your idea of having the word "Stage" in stage types.  I already have Event in event types (except for one).  Adding "Op" also makes sense.  I'll do that in a separate commit.

---
In reply to: [293455914](https://github.com/Azure/azure-iot-sdk-python-preview/pull/106#discussion_r293455914) [](ancestors = 293455914)